clear-intermediate-datasets/python-runnables/clear_int_datasets/runnable.py

```diff
@@ -0,0 +1,129 @@
+import dataiku
+import logging
+
+from dataiku.runnables import Runnable, ResultTable
+
+logging.basicConfig(format="%(asctime)s [%(levelname)s] %(message)s", level=logging.INFO)
+logging.getLogger().setLevel(logging.INFO)
+
+def populate_result_table_with_list():
+    # TODO
+    # Populate result table from list with arbitrary table size
+    return NotImplementedError()
+
+
+class MyRunnable(Runnable):
+    """The base interface for a Python runnable."""
+
+    def __init__(self, project_key, config, plugin_config):
+        """
+        :param project_key: the project in which the runnable executes
+        :param config: the dict of the configuration of the object
+        :param plugin_config: contains the plugin settings
+        """
+
+        self.project_key = project_key
+        self.config = config
+        self.plugin_config = plugin_config
+
+    def get_progress_target(self):
+        """
+        If the runnable will return some progress info, have this function return a tuple of 
+        (target, unit) where unit is one of: SIZE, FILES, RECORDS, NONE
+        """
+
+        return None
+
+    def run(self, progress_callback):
+
+        # Retrieve macro parameters:
+        is_dry_run = self.config.get("is_dry_run")
+        keep_partitioned = self.config.get("keep_partitioned")
+        keep_shared = self.config.get("keep_shared")
+        logging.info("DRY RUN is set to {}".format(str(is_dry_run)))
+        logging.info("KEEP PARTITIONED is set to {}".format(str(keep_partitioned)))
+        logging.info("KEEP SHARED is set to {}".format(str(keep_shared)))
+
+        # Initialize macro result table:
+        result_table = ResultTable()
+        result_table.add_column("dataset", "dataset", "STRING")
+        result_table.add_column("status", "status", "STRING")
+
+        client = dataiku.api_client()
+        project = client.get_project(self.project_key)
+        all_datasets = project.list_datasets()
+        all_recipes = project.list_recipes()
+
+        # Build deduplicated lists of input/output datasets:
+        input_datasets = []
+        output_datasets = []
+        for rcp in all_recipes:
+            # List dataset names that are inputs:
+            rcp_in = [x["ref"] for x in rcp["inputs"]["main"]["items"]]
```

Comment:
This does not work if in the flow we have either: 
- Plugin recipes
- Code recipes (python/shell/...) with no inputs.

Reply:
@du-phan the breaking case with plugin recipes happens with said recipes anywhere in the flow or just at the beginning?